### PR TITLE
feat: New polling logic for DE

### PR DESCRIPTION
**Pull request by arjunaj5**
## Clickup
https://app.clickup.com/t/86cwh9db4

## Code Coverage
Please add code coverage here

## UI Preview
**[Slack Thread containing recording](https://fylein.slack.com/archives/C07FT7RSXJR/p1726554119032609?thread_ts=1726553321.559229&cid=C07FT7RSXJR)**

### CODE REVIEW

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 1]
```diff
@@ -445,6 +452,97 @@ export class MyExpensesPage implements OnInit {
...
+
+    const hasUserManuallyEnteredData =
+      isZeroAmountPerDiemOrMileage ||
+      ((expense.amount || expense.claim_amount) && isNumber(expense.amount || expense.claim_amount));
+    const isDataExtracted = !!expense.extracted_data;
+
+    // this is to prevent the scan failed from being shown from an indefinite amount of time.
+    const hasScanExpired = expense.created_at && dayjs(expense.created_at).diff(Date.now(), 'day') < 0;
+    return !!(hasUserManuallyEnteredData || isDataExtracted || hasScanExpired);
+  }
```
Comment (arjunaj5) [inline src/app/fyle/my-expenses/my-expenses.page.ts:478]: taken from expenses card component

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 2]
```diff
@@ -445,6 +452,97 @@ export class MyExpensesPage implements OnInit {
...
+   * @param {PlatformExpense[]} initialExpenses - The initial list of expenses.
+   * @returns {Observable<PlatformExpense[]>} - Observable that emits updated expenses.
+   */
+  private pollIncompleteExpenses(
+    incompleteExpenseIds: string[],
+    expenses: PlatformExpense[]
+  ): Observable<PlatformExpense[]> {
+    let updatedExpensesList = expenses;
+    // Create a stop signal that emits after 30 seconds
+    const stopPolling$ = timer(30000);
```
Comment (arjunaj5) [inline src/app/fyle/my-expenses/my-expenses.page.ts:528]: taking 30 seconds as the limit for polling after every 5 seconds.

File: src/app/shared/components/expenses-card-v2/expenses-card.component.ts [hunk 3]
```diff
@@ -216,7 +216,7 @@ export class ExpensesCardComponent implements OnInit {
           (that.homeCurrency === 'USD' || that.homeCurrency === 'INR')
         ) {
           that.isScanCompleted = that.checkIfScanIsCompleted();
-          that.isScanInProgress = !that.isScanCompleted;
+          that.isScanInProgress = !that.isScanCompleted && !this.expense.extracted_data;
```
Comment (arjunaj5) [inline src/app/shared/components/expenses-card-v2/expenses-card.component.ts:219]: if `isScanInProgress` is false and `isScanCompleted` is also false, then the scan failed message is shown. This is the current behaviour.

**Review by Chethan-Fyle — APPROVED** [review]

File: src/app/shared/components/expenses-card-v2/expenses-card.component.ts (hunk 3 above)
Comment (mvaishnavi) [reply src/app/shared/components/expenses-card-v2/expenses-card.component.ts:219]: should this be `that.expense.extracted_data`?

File: src/app/shared/components/expenses-card/expenses-card.component.ts [hunk 4]
```diff
@@ -208,7 +208,7 @@ export class ExpensesCardComponent implements OnInit {
           (that.homeCurrency === 'USD' || that.homeCurrency === 'INR')
         ) {
           that.isScanCompleted = that.checkIfScanIsCompleted();
-          that.isScanInProgress = !that.isScanCompleted;
+          that.isScanInProgress = !that.isScanCompleted && !this.expense.tx_extracted_data;
```
Comment (mvaishnavi) [inline src/app/shared/components/expenses-card/expenses-card.component.ts:211]: same as above.

File: src/app/shared/components/expenses-card-v2/expenses-card.component.ts (hunk 3 above)
Comment (arjunaj5) [reply src/app/shared/components/expenses-card-v2/expenses-card.component.ts:219]: no, If extracted_data is present, then it means that the scan is completed.

**Pull request merged**